sui_programmability/framework/sources/Governance/ValidatorSet.move

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module Sui::ValidatorSet {
+    use Std::Vector;
+
+    use Sui::Coin::Coin;
+    use Sui::SUI::SUI;
+    use Sui::Transfer;
+    use Sui::TxContext::{Self, TxContext};
+    use Sui::Validator::{Self, Validator};
+
+    friend Sui::SuiSystem;
+
+    const EDUPLICATE_VALIDATOR: u64 = 0;
+
+    const EINVALID_VALIDATOR_INDEX: u64 = 1;
+
+    const EVALIDATOR_NOT_FOUND: u64 = 2;
+
+    const EDUPLICATE_WITHDRAW: u64 = 3;
+
+    const ENOT_ENOUGH_VALIDATORS: u64 = 4;
+
+    struct ValidatorSet has store {
+        /// The current list of active validators.
+        /// The stake amount is sorted, with highest stake first.
+        active_validators: vector<Validator>,
+
+        /// List of new validator candidates added during the current epoch.
+        /// They will be processed at the end of the epoch.
+        pending_validators: vector<Validator>,
+
+        /// Withdraw requests from the validators. Each element is an index
+        /// pointing to `active_validators`.
+        pending_withdraws: vector<u64>,
+    }
+
+    public(friend) fun new(init_active_validators: vector<Validator>): ValidatorSet {
+        ValidatorSet {
+            active_validators: init_active_validators,
+            pending_validators: Vector::empty(),
+            pending_withdraws: Vector::empty(),
+        }
+    }
+
+    /// Called by `SuiSystem`, add a new validator to `pending_validators`, which will be
+    /// processed at the end of epoch.
+    public(friend) fun request_add_validator(self: &mut ValidatorSet, validator: Validator) {
+        assert!(
+            contains_duplicate_validator(&self.active_validators, &validator)
+                && contains_duplicate_validator(&self.pending_validators, &validator),
+            EDUPLICATE_VALIDATOR
+        );
+        Vector::push_back(&mut self.pending_validators, validator);
+    }
+
+    /// Called by `SuiSystem`, to withdraw a validator.
+    /// If the validator to be withdrawn is a pending one, it's removed immediately and returned
+    /// to the validator's sui_address; otherwise the index is added to `pending_withdraws` and
+    /// will be processed at the end of epoch.
+    public(friend) fun request_withdraw_validator(
+        self: &mut ValidatorSet,
+        validator_address: address,
```

Comment:
Nit: I think it would be slightly safer/clearer to take `TxContext` as input + compute `sender` here instead of doing it in the caller. Similar elsewhere for `request_add_stake` etc.